ops/main.py

```diff
@@ -42,6 +43,134 @@
+    # The unit that is departing a relation, for example 'remote/42' (from JUJU_DEPARTING_UNIT).
+    relation_departing_unit_name: Optional[str] = None
+
+    # The name of the relation, for example 'database' (from JUJU_RELATION).
+    relation_name: Optional[str] = None
+
+    # The id of the relation,
+    # for example 1 (integer) if the original environment variable's value is 'database:1'
+    # (from JUJU_RELATION_ID).
+    relation_id: Optional[int] = None
+
+    # The name of the remote app, for example 'remoteapp1' (from JUJU_REMOTE_APP).
+    remote_app: Optional[str] = None
+
+    # The name of the remote unit, for example 'remoteapp1/0' (from JUJU_REMOTE_UNIT).
+    remote_unit_name: Optional[str] = None
+
+    # The ID of the secret,
+    # for example 'secret:dcc7aa9c-7202-4da6-8d5f-0fbbaa4e1a41' (from JUJU_SECRET_ID).
+    secret_id: Optional[str] = None
+
+    # The label of the secret, for example 'db-password' (from JUJU_SECRET_LABEL).
+    secret_label: Optional[str] = None
+
+    # The revision of the secret, for example 42 (integer) (from JUJU_SECRET_REVISION).
+    secret_revision: Optional[int] = None
+
+    # The storage name, for example 'my-storage'
+    # if the original environment variable's value is 'my-storage/1' (from JUJU_STORAGE_ID).
+    storage_name: Optional[str] = None
+
+    # The name of the unit, for example 'myapp/0' (from JUJU_UNIT_NAME)
+    unit_name: str = ''
+
+    # The version of Juju, for example '3.4.0' (from JUJU_VERSION)
+    version: Optional[JujuVersion] = None
+
+    # The name of the workload, for example 'workload' (from JUJU_WORKLOAD_NAME).
+    workload_name: Optional[str] = None
+
+    @classmethod
+    def from_dict(cls, env: Mapping[str, Any]) -> '_JujuContext':
+        return _JujuContext(
+            action_name=env.get('JUJU_ACTION_NAME') or None,
+            action_uuid=env.get('JUJU_ACTION_UUID') or None,
+            charm_dir=(
+                Path(f'{__file__}/../../..').resolve()
+                if not env.get('JUJU_CHARM_DIR')
+                else Path(env.get('JUJU_CHARM_DIR', '')).resolve()
+            ),
+            debug=env.get('JUJU_DEBUG', '').lower() == 'true',
+            dispatch_path=env.get('JUJU_DISPATCH_PATH', ''),
+            model_name=env.get('JUJU_MODEL_NAME', ''),
+            model_uuid=env.get('JUJU_MODEL_UUID', ''),
+            notice_id=env.get('JUJU_NOTICE_ID') or None,
+            notice_key=env.get('JUJU_NOTICE_KEY') or None,
+            notice_type=env.get('JUJU_NOTICE_TYPE') or None,
+            pebble_check_name=env.get('JUJU_PEBBLE_CHECK_NAME') or None,
+            relation_departing_unit_name=env.get('JUJU_DEPARTING_UNIT') or None,
+            relation_name=env.get('JUJU_RELATION') or None,
+            relation_id=(
+                int(env['JUJU_RELATION_ID'].split(':')[-1])
+                if env.get('JUJU_RELATION_ID')
+                else None
+            ),
+            remote_app=env.get('JUJU_REMOTE_APP') or None,
```

Comment:
`env.get(...) or None` is redundant since `dict.get` will return None is there's no key.

Reply:
It mainly handles the corner case where the key is set to an empty str, I learned this trick from Ben :)

```python
>>> import os
>>> os.environ['foo'] = ''
>>> foo = os.getenv('foo')
>>> foo
''
>>> assert foo is None
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

Reply:
Oh, makes sense then :)